Created blog post on how to swap out components in Kubernetes apps (#4218)
pulumi/docs: content/blog/switching-the-application-stack-from-pern-to-mern/index.md

---
title: "Switching the application stack from PERN to MERN"
date: 2020-09-18
meta_desc: Demonstrating the simplicity, modularity, and reusability of running an application on Kubernetes using Pulumi.
meta_image: meta.png
authors: ["vova-ivanov"]
tags: ["aws", "typescript", "docker", "kubernetes"]
---

> In this blog post we will return to the PERN application we previously [migrated to Kubernetes]({{< relref "/blog/deploying-a-pern-stack-application-to-aws" >}}), and switch out the PostgreSQL database for MongoDB. Although at first glance it might seem like a difficult task, the straigthforward design of Pulumi and Kubernetes allows us to easily transition the application form a PERN stack to a MERN one.

Reviewer: ```suggestion
In this blog post, we return to the PERN application we previously [migrated to Kubernetes]({{< relref "/blog/deploying-a-pern-stack-application-to-aws" >}}) and replace the PostgreSQL database with MongoDB. Although it might seem like a difficult task initially, the straightforward design of Pulumi and Kubernetes allows us to easily transition the application form a PERN stack to a MERN one.
```

<!--more-->

> The main difference between PostgreSQL and MongoDB is data is stored. PostgreSQL is a relational database that allows users to create complex tables and relations between values. In contrast, MongoDB is a key-value database that stores data in simple collections without any special relations. Because the application takes the form of three distinct components--the client, server, and database--almost all of the changes will occur in the database component. In fact, we can reuse most of the existing code and will be able to change the database without additional hassle.

Reviewer: ```suggestion
The main difference between PostgreSQL and MongoDB is data is stored. PostgreSQL is a relational database that allows users to create complex tables and relations between values. In contrast, MongoDB is a key-value database that stores data in simple collections instead of a relational model. Because the application takes the form of three distinct components--the client, server, and database--almost all of the changes will occur in the database component. In fact, we can reuse most of the existing code and will be able to change the database without additional hassle.
```

> The main difference between PostgreSQL and MongoDB is data is stored. PostgreSQL is a relational database that allows users to create complex tables and relations between values. In contrast, MongoDB is a key-value database that stores data in simple collections without any special relations. Because the application takes the form of three distinct components--the client, server, and database--almost all of the changes will occur in the database component. In fact, we can reuse most of the existing code and will be able to change the database without additional hassle.

Reviewer: ```suggestion
The main difference between PostgreSQL and MongoDB is data is stored. PostgreSQL is a relational database that allows users to create complex tables and relations between values. In contrast, MongoDB is a key-value database that stores data in simple collections instead of a relational model. Because the application takes the form of three distinct components--the client, server, and database--almost all of the changes will occur in the database component. In fact, we can reuse most of the existing code and will be able to change the database without additional hassle.
```

> The main difference between PostgreSQL and MongoDB is the way they store their data: PostgreSQL is a relational database which allows users to create complex tables and relations between values, while MongoDB is a key-value database that stores data in simple collections without any special relations. Because the application takes the form of three distinct components--the client, server, and database--almost all of the changes will occur in the database component. We can in fact reuse most of the existing code, and will be able to change the database without additional hassle.

Reviewer: ```suggestion
The main difference between PostgreSQL and MongoDB is data is stored. PostgreSQL is a relational database that allows users to create complex tables and relations between values. In contrast, MongoDB is a key-value database that stores data in simple collections without any special relations. Because the application takes the form of three distinct components--the client, server, and database--almost all of the changes will occur in the database component. In fact, we can reuse most of the existing code and will be able to change the database without additional hassle.
```

```bash
$ mkdir examples && cd examples
$ git init
$ git remote add origin -f https://github.com/pulumi/examples/
$ git config core.sparseCheckout true
$ echo aws-ts-k8s-voting-app >> .git/info/sparse-checkout
$ git pull origin master
```

The next step is to create a new directory and initialize a Pulumi project with `pulumi new aws-typescript`.

```bash
$ mkdir aws-ts-k8s-mern-voting-app && cd aws-ts-k8s-mern-voting-app
$ pulumi new aws-typescript
```

The project will require several configuration variables, which we set using `pulumi config set`. The variables are used to configure the MongoDB user account, initializing the database schema and table, and its region.

```bash
$ pulumi config set sql-user-name <NAME>
$ pulumi config set sql-user-password <PASSWORD> --secret
$ pulumi config set aws:region <REGION>
```

The `package.json` file lists the libraries used by the project. We will add the following to the `dependencies` section:

```json
    "@pulumi/eks": "^0.20.0",
    "@pulumi/cloud-aws": "^0.19.0"
```

> As most of the components in our project are nearly identical to those in the PERN Kubernetes application, we can copy many of them into our directory.

Reviewer: ```suggestion
As most of our project components are nearly identical to those in the PERN Kubernetes application, we can copy many of them into our directory.
```

EXPOSE 5432

RUN apt update && \
    apt install -y mongodb

ADD database /database

CMD [ "/database/startDatabase.sh" ]
```

Our setup doesn't require any files, and so the only thing we need to put in the `database` folder is a `startDatabase.sh` script that launches the database, configures users, and creates a table.

```bash
#!/bin/bash
set -exu
FILE=/persistentVolume/mongod.lock

if test -f "$FILE"; then
    echo "/persistentVolume already contains MongoDB, no need to initialize database."
    mongod --fork --dbpath /persistentVolume/ --bind_ip_all --logpath /persistentVolume/mongoLogs
else
    echo "/persistentVolume is empty, and we need to initialize the MongoDB database."
    mongod --fork --dbpath /persistentVolume/ --bind_ip_all --logpath /persistentVolume/mongoLogs
    echo "use $DATABASE_NAME
    db.choices.insert({ _id: 0, text: \"Tabs\", vote_count: 0 })
    db.choices.insert({ _id: 1, text: \"Spaces\", vote_count: 0 })
    db.createUser(
        {
            user: \"$USER_NAME\",
            pwd: \"$USER_PASSWORD\",
            roles: [ { role: \"readWrite\", db: \"$DATABASE_NAME\" } ]
        }
    )" | mongo
fi

while true; do
    sleep 3600
done
```

> With the database in place, all that is left is to make some small MongoDB-specific changes to the way queries are sent by the client and server components.

Reviewer: ```suggestion
With the database in place, all that is left is to make some small MongoDB-specific changes to how the client and server components send queries.
```

> The server component will be changed to use a `model.js` file, which acts like a blueprint for the choices that the database has.

Reviewer: ```suggestion
The server component will be changed to use a model.js file, which acts as a blueprint for the database's choices.
```

```javascript
import React, { Fragment, useEffect, useState } from "react";
const fullServerUrl = "http://" + window.SERVER_URL + ":5000/voting";

const ListChoices = () => {
  const [choices, setChoices] = useState([]);

  const getChoices = async () => {...};

  useEffect(() => {...}, []);

  return (
    <Fragment>
      <h1 className="text-center mt-5">Pulumi Voting App</h1>
      {" "}
      <table className="table mt-5 text-center">
        <thead>
          <tr>
            <th>Choice</th>
            <th>Votes</th>
          </tr>
        </thead>
        <tbody>
          {choices.map(choice => (
            <tr key={choice._id}>
              <td>{choice.text}</td>
              <td>{choice.vote_count}</td>
              <td>
                <button
                  className="btn btn-success"
                  onClick={() => castVote(choice._id)}
                >
                  Vote
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </Fragment>
  );
};

async function castVote(id) {...}

export default ListChoices;
```

> Now that, server, client, and database have been fully converted to use MongoDB, we can make the last remanining change in the infrastructure by updating the database deployment and service.

Reviewer: ```suggestion
Now that, server, client, and database have been fully converted to use MongoDB, we can make the last remaining change in the infrastructure by updating the database deployment and service.
```

                        { name: "USER_PASSWORD", value: sqlUserPassword },
                        { name: "MONGODB_ADDRESS", value: mongodbAddress },
                        { name: "MONGODB_PORT", value: String(27017) },
                        { name: "DATABASE_NAME", value: databaseName },
                    ],
                    resources: {
                        limits: {
                            memory: "500Mi",
                            cpu: "500m"
                        }
                    }
                }],
            }
        }
    }}, {
    provider: eksCluster.provider,
});

const serversideListener = new k8s.core.v1.Service(...);

const clientAppName = ...;
const clientAppLabels = { ... };
const clientDeployment = new k8s.apps.v1.Deployment(...);

const clientsideListener = new k8s.core.v1.Service(...);
```

To give us a simple way to connect to our application, we export the clientside listener's address. We can now open a browser window with the URL and port number to view our application.

```typescript
export const kubeConfig = eksCluster.kubeconfig;
export const URL = clientsideListener.status.loadBalancer.ingress[0].hostname;
```

> In this example, we showed how to change a PERN stack based application to use MERN, and demonstrated the ease with which a designer can swap out different components within their application. The isolated nature of containerized applications allows for a great deal of modularity, and can significantly reduce the amount of time spent renovating previously written code to fit any new components that are added.

Reviewer: ```suggestion
In this example, we showed how to change a PERN stack-based application to use MERN, and demonstrated the ease with which a designer can swap out different components. The isolated nature of containerized applications allows for a great deal of modularity. It can significantly reduce the amount of time spent renovating previously written code to fit any added new components.
```

> The blog post's code can be [found on Github](https://github.com/pulumi/examples/tree/master/aws-ts-k8s-mern-voting-app).

Reviewer: ```suggestion
The MERN application code is [on Github](https://github.com/pulumi/examples/tree/master/aws-ts-k8s-mern-voting-app).
```
